doc/event-alarm-framework/events-producer.md

```diff
@@ -0,0 +1,212 @@
+- Watch syslog messages for specific pattern; Upon match parse the required data out and call event reporter.
+- Any other means of getting alerted on an event
+
+#### Log message based detection
+At high level:
+1. This is a two step process.
+2. The process raising the event sends a sylog message out.
+3. A watcher scans all the syslog messages emitted and parse/check for events of interest.
+4. When matching message arrives, raise the event
+
+Here you have code that sends the log and a watcher who has the regex pattern for that log message to match. Anytime the log messsage is changed the pattern has to be updated for the event to fire consistently across releases.
+
+Though this sounds like a redundant/roundabout way, this helps as below.
+- For III party code, not owned by SONiC, this is an acceptable solution
+- For code that SONiC own, a code update would take in the order of months to reach thousands of switches, but this approach can be applied instantly to every switch
+- A new event addition could be as simple as copying couple of small files to a switch and a rsyslog/container restart.
+
+##### Design at high level
+- Configure a rsyslog plugin as per process with rsyslog.d.
+- For logs raised by host processes, configure this plugin at host.
+- For logs raised by processes inside the container, configure for rsyslog.d running inside the container.
+- In this mode, there will be plugin processes running in host & in containers as one instance per process.
+- Provide the regex patterns to use for matching events as i/p to the plugin (*list of patterns for a process*).
+- Each plugin scans messasges **only** from a *single* process and matches only against patterns for that process.
+- For messages that match a pattern, retrieve parameters of interest and fire event using event reporter.
+- The rsyslog plugin binary, which does the parsing & reporting is a single binary in host, shared/used by all plugins.
+- The rsyslog plugin binary being under host control, ensures a single/unified behavior across all.
+- The unit tests can use hardcoded log messages to validate regex.
+
+
+![image](https://user-images.githubusercontent.com/47282725/157343412-6c4a6519-c27b-459b-896b-7875d8f952b8.png)
+
+##### Pros & Cons
+
+###### Pro
+1) This support is external to app, hence adapts well with III party applications, bgp, teamd, ... 
+2) This feature can be added to released builds too, as all it takes is to copy two files into each container and restart rsyslogd in the container
+3) The regex for parsing being local to container, it supports any container upgrade transaparently.
+4) The rsyslog plugin binary is maintained by host, hence provide a overall control across all containers.
+5) The message parsing load is distributed as per container. Within a container parsing is done at the granularity of per process with no extra cost as rsyslogd already pre-parsed it per-process, *always*
```

Comment:
>rsyslogd already pre-parsed it per-process [](http://example.com/codeflow?start=152&length=42)

Do you have reference link to explain "rsyslogd already pre-parsed it per-process"?

Reply:
It is one of the properties for use as provided by rsyslog.d in its .conf.